File: src/bin/rad.pt/trad_can/exp_int3.f90
!****************************************************************************************
!                               FUNCTION EXP_INT3
!****************************************************************************************
! Calculates the 3rd exponential integral from "Numerical Recipes in Fortran 90"
! Press, Teukolsky et al

DOUBLE PRECISION function exp_int3(tau)

USE parameters
implicit none
REAL (KIND=DBL), INTENT (IN) :: tau
REAL (KIND=DBL) :: b, c, d, h, a, del, eps, big, fact, psi
INTEGER :: l_, k_



   ! Set some function parameters
   ! 1. Desired relative error. Not smaller than machine precision.
        eps = EPSILON (tau)
   ! 2. Number near the largest representable floating-point number
        big = HUGE (tau) * eps


   if (tau < 0d0) then
       ! Optical depth less than zero
       write (*,*) 'Negative Optical depth in exp_int3. Program stopped.'
       stop
   else if (tau == 0d0) then
       ! Special case
       exp_int3 = 1d0 / (THIRD - 1)
   else if (tau > 1d0) then
       ! Uses Lentz's algorithm
       b = tau + THIRD
       c = big
       d = 1d0 / b

      if (abs(b) < TOL_ZERO) then
         write (*,*) 'zero error in ' , 'expint 1'
         stop
      end if


       h = d
       do k_ = 1, MAX_ITER
           a = - REAL(k_ * (THIRD - 1 + k_))
           b = b + 2d0
           d = 1d0 / ( a * d + b)

      if (abs(a* d + b) < TOL_ZERO) then
         write (*,*) 'zero error in ' , 'expint 2'
         stop
      end if

      if (abs(c) < TOL_ZERO) then
         write (*,*) 'zero error in ' , 'expint 3'
         stop
      end if


           c = b + a / c
           del = c * d
           h = h * del
           if (ABS(del - 1d0) <= eps) EXIT
       end do
       if (k_ > MAX_ITER) then           ! Safety condition
           write (*,*) 'exp_int3: continued fraction failed'
           stop
       end if
       exp_int3 = h * EXP(- tau)
   else
       ! Convergence is too slow for 0 < tau < 1 .
       ! Use series representation instead.
       exp_int3 = 1d0 / real(THIRD-1)
       fact = 1d0
       do k_ = 1, MAX_ITER
           fact = - fact * tau / REAL(k_)
           if (k_ /= THIRD-1) then
               del = - fact / real(k_ - (THIRD - 1))
           else
               ! Compute psi
               psi = - EULER
               do l_ = 1, THIRD-1
                   psi = psi + 1d0 / REAL(l_)
               end do
               del = fact * ( - LOG(tau) + psi)
           end if
           exp_int3 = exp_int3 + del
           if (ABS(del) < ABS(exp_int3) * eps) EXIT
       end do
       if (k_ > MAX_ITER) then        ! Safety condition
           write (*,*) 'exp_int3: series failed'
           stop
       end if
   end if


   return

end function exp_int3

File: src/bin/rad.pt/solar_can/exp_int3.c
/****************************************************************************************
                                FUNCTION EXP_INT3
  Calculates the 3rd exponential integral from "Numerical Recipes in Fortran 90"
  Press, Teukolsky et al; originally coded by M. Tribbeck; recoded in C by D. Marks
  NWRC, Jul 2002
*****************************************************************************************/

#include <math.h>
#include "ipw.h"
#include "pgm.h"

#define THIRD		3
#define MAX_ITER	100
#define EULER		0.5772
#define EPSILON		0.0001
#define NO_CANOPY	(1.0 / (double) (THIRD - 1.0))


double
exp_int3 (
	double	tau	/* optical depth (lai / 2) */)
{
	double	a;
	double	b;
	double	c;
	double	d;
	double	h;
	double	big;
	double	del;
	double	eps;
	double	fact;
	double	psi;
	double	no_canopy;
	double	return_val;

	int	l;
	int	k;

	/*	set no canopy value	*/
	no_canopy = NO_CANOPY;

	/*	set integration limits	*/

	/** don't really understand the f90 functions
	 ** EPSILON(x) and HUGE(x), and can't figure why:
	 ** "eps" is the convergence limit and is set
	 ** to the constant EPSILON (above) but big is confusing...
	 ** if 
	 **	eps = EPSILON(tau)
	 ** then:
	 **	big = HUGE(x) * eps	
	 ** but I'll do it that way...
	 ** I need to figure out a "BIG" value that is reasonable
	 **/

	eps = DBL_EPSILON;
	big = DBL_MAX * eps;

	if (tau < 0.0) {
		/*	bad argument passed	*/
		usrerr("exp_int3: negative canopy optical depth passed");
		return (ERROR);
	}
	else if (tau == 0.0) {
		/*	no canopy, return 0.5	*/
		return (no_canopy);
	}
	else if (tau > 1.0) {
		/*	valid canopy optical depth, greater than 1.0	*/
		/*	use Lentz's algorithm	*/
		b = tau + THIRD;
		c = big;
		d = 1.0 / b;

		h = d;
		k = 1;
		while (k <= MAX_ITER) {
			a = - (double) (k * (THIRD - 1 + k));
			b += 2.0;
			d = 1.0 / ((a * d) + b);
			c = b + (a / c);
			del = c * d;
			h = h * del;
			if (fabs(del - 1.0) <= eps)
				break;
			k++;
		}
		if (k > MAX_ITER)
			error("convergence failed: tau = %g", tau);
		return_val = h * exp(-tau);
		return (return_val);
	}
	else {
		/*	valid optical depth, between 0 and 1	*/
		/*	series representation used to speed convergence	*/
		return_val = no_canopy;
		fact = 1.0;
		k = 1;
		while (k <= MAX_ITER) {
			fact = - fact * tau / (double) k;
			if (k != (THIRD - 1))
				del = - fact / (double) (k - (THIRD - 1));
			else {
				psi = - EULER;
				l = 1;
				while (l <= (THIRD - 1)) {
					psi += 1.0 / (double) l;
					l++;
				}
				del = fact * ( - log(tau) + psi);
			}
			return_val = return_val + del;
			if (fabs(del) < (fabs(return_val) * eps))
				break;
			k++;
		}
		if (k > MAX_ITER)
			error("convergence failed: tau = %g", tau);
		return (return_val);
	}
}
